File: Introduction_to_Robotics_Course/Chapter14/Lesson2/Chapter14_Lesson2.cpp
#include <iostream>
#include <cmath>

double human_command(double time) {
    if (time < 5.0) {
        return 1.0;
    } else {
        return -1.0;
    }
}

int main() {
    const double T = 0.01;
    const double T_end = 10.0;
    const double k_gain = 1.0;
    const double alpha = 0.4;
    const double U_max = 2.0;

    const int N = static_cast<int>(T_end / T);
    double x = 0.0;
    double t = 0.0;

    for (int k = 0; k < N; ++k) {
        double u_h = human_command(t);
        double u_a = -k_gain * x;
        double u = alpha * u_h + (1.0 - alpha) * u_a;

        // Saturation
        if (u > U_max) u = U_max;
        if (u < -U_max) u = -U_max;

        x = x + T * u;
        t = t + T;
    }

    std::cout << "Final position: " << x << std::endl;
    return 0;
}
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter14/Lesson2/Chapter14_Lesson2.m
T = 0.01;
T_end = 10.0;
k_gain = 1.0;
alpha = 0.4;
U_max = 2.0;

N = floor(T_end / T);
t = linspace(0, T_end, N + 1);
x = zeros(1, N + 1);

human_command = @(time) (time < 5.0) * 1.0 + (time >= 5.0) * (-1.0);

for k = 1:N
    u_h = human_command(t(k));
    u_a = -k_gain * x(k);
    u = alpha * u_h + (1 - alpha) * u_a;
    % Saturation
    u = max(-U_max, min(U_max, u));
    x(k + 1) = x(k) + T * u;
end

plot(t, x);
xlabel('Time [s]');
ylabel('Position x(t)');
title('1-DOF Shared Autonomy Simulation');
grid on;
      
